add php.ini entry to set default user_agent for curl

paresy opened the pull request:
> https://www.php.net/manual/en/filesystem.configuration.php#ini.user-agent
> Define the user agent for PHP to send.
> 
> https://www.php.net/manual/en/soapclient.construct.php
> The user_agent option specifies string to use in User-Agent header.

In the same spirit i'd like to propose a php.ini entry to set a default user agent for all cURL requests.

The patch seems very straightforward and i am not sure if it requires a full RFC process.
If everything is fine i would also send a PR to update the docs respectively.

PS: This was my first test to write for PHP and the experience was good 👍

carusogabriel commented:
Hey @paresy, thanks for your proposal!

Would you mind sending a message to internals@lists.php.net (check http://externals.io and https://news-web.php.net/php.internals) telling everyone about this proposal?

awslabspl reviewed (commented):
Lgtm

paresy commented:
So far there hasn't been a lot of responses. https://www.mail-archive.com/internals@lists.php.net/msg106186.html
The one i got preferred the proposed solution instead of a broader/wider approach.

How can we proceed? Shall i bump the topic on the mailing list? Or shall i get a PR done for the docs?

TysonAndre commented:
nit: Documentation of what the new ini setting does and what the default value is should be added to php-src/php.ini-development and php-src/php.ini-production for this proposal


```
[curl]
; A default value for the CURLOPT_CAINFO option. This is required to be an
; absolute path.
;curl.cainfo =
```

> How can we proceed? Shall i bump the topic on the mailing list? Or shall i get a PR done for the docs?

Docs outside of php-src and the RFC document are generally only done after an RFC is accepted and merged, I think

https://wiki.php.net/rfc/howto

> Listen to the feedback, and try to answer/resolve all questions. Update your RFC to document all the issues and discussions. Cover both the positive and negative arguments. Put the RFC URL into all your replies.
**When discussion ends, and a minimum period of two weeks has passed since you mailed internals@lists.php.net in step 4, consider one day heads up mail on the mailing list and then you can move your RFC to “Voting” status.** There should be no open questions in the RFC.

ramsey commented:
@TysonAndre Are you recommending that @paresy create an RFC for `curl.user_agent`?

paresy commented:
Sorry for letting this slip so long. I updated the PR to include examples inside the php.ini files. I also tried to clarify and cross-link both options.

I also added a quick link to "https://php.net/curl-user-agent". I am not sure how to add this to the documentation, but if it is possible to do so, i will add it after this gets merged :)

Thanks for looking into it!

TysonAndre commented on php.ini-development:
```diff
@@ -1919,6 +1920,11 @@ ldap.max_links = -1
 
+; Define the User-Agent when using cURL for requests. PHP's default setting for this is empty.
+; See user_agent (without prefix) to set the User-Agent when using the HTTP wrapper (e.g. file_get_contents)
+; https://php.net/curl-user-agent
```
```suggestion
```

I don't see a need to include this link - `curl.cainfo` doesn't have a link either, and the link might not ever be created.

TysonAndre commented on php.ini-development:
```diff
@@ -1919,6 +1920,11 @@ ldap.max_links = -1
 ; absolute path.
 ;curl.cainfo =
 
+; Define the User-Agent when using cURL for requests. PHP's default setting for this is empty.
```
```suggestion
; Define the User-Agent when using cURL for requests. When this is empty or unset, curl's default user agent is used instead.
```

TysonAndre reviewed (commented):
> Are you recommending that @paresy create an RFC for `curl.user_agent`?

I'd recommend creating an RFC - both of the responses on the rfc thread were slightly negative (recommended alternatives). https://externals.io/message/114001#114039 - overall, I can't tell if that's representative as voters as a whole because there's not much to say about a simple change if a voter assumed the author was going to start an RFC

I'm slightly in favor of this, e.g. if there's third party libraries or applications that use curl directly and don't set a user agent (or would be too much effort to bother refactoring), it might be useful to include information about the host in question (app type/tier name/versions, etc) in the default curl user agent in php.ini - it'd save a bit of time debugging if the original application was hidden by a load balancer.

TysonAndre commented on ext/curl/interface.c:
```diff
@@ -232,6 +232,7 @@ static inline int build_mime_structure_from_hash(php_curl *ch, zval *zpostfields
 /* {{{ PHP_INI_BEGIN */
 PHP_INI_BEGIN()
 	PHP_INI_ENTRY("curl.cainfo", "", PHP_INI_SYSTEM, NULL)
+	PHP_INI_ENTRY("curl.user_agent", "", PHP_INI_ALL, NULL)
```
After thinking about it more, I'd personally be in favor of PHP_INI_SYSTEM but opposed to the permissive PHP_INI_ALL. PHP_INI_SYSTEM would forbid calling ini_set at runtime, but still allow system configuration.

See https://www.php.net/manual/en/configuration.changes.modes.php and https://www.php.net/manual/en/configuration.changes.php (e.g. for web servers that use apache, the setting can be set in php.ini, and/or overridden for individual apache virtual hosts in /etc/httpd/vhosts.d/ with `php_value`) 

`php -d curl.user_agent='Some user agent' myCLIScript.php` would also always work if different CLI applications needed different user agents

-------

The reason I think the ability to call ini_set at runtime might be a problem:
1. If an application has two libraries that use the default curl user agent (libA and libB), and LibA were to call `ini_set('curl.user_agent', 'LibA/1.2.3')` and not bother to restore the user agent (or throw, etc), then the application would unexpectedly and spuriously sometimes call the endpoints that libB calls with the wrong user agent of LibA.

   Async frameworks using generators and php 8.1's fibers would make that a larger concern
2. So the application using LibB calling ServerB with a user agent from LibA may cause issues, such as developers wasting time debugging because of being mislead about what the service's caller actually is, or ServerB parsing the wrong versions from the user agent, or ServerB rejecting requests because the user agent is unrecognized or forbidden



Obviously, for `file_get_contents`, there's no way to override the user agent, so it's useful that `user_agent` can be overridden at runtime. But curl_setopt can already override the user agent

paresy replied:
Sorry for the late response. As far as i have checked you can override the user_agent during runtime using file_get_contents. Therefore i wanted to keep curl.user_agent in the same mode as the original user_agent.

This is my code to verify this:
```
<?php

ini_set('user_agent', 'test');

$context = stream_context_create([
  'http' => [
    'user_agent' => 'hello'
  ]
]);

$body = file_get_contents("https://www.whatsmyua.info/api/v1/ua", false, $conte$
$json = json_decode($body, true);

echo $json[0]["ua"]["rawUa"];
```

TysonAndre replied:
> Sorry for the late response. As far as i have checked you can override the user_agent during runtime using file_get_contents. Therefore i wanted to keep curl.user_agent in the same mode as the original user_agent.

That only affects the call in question - I'm talking about ini_set affecting completely unrelated libraries that don't share a context

paresy replied:
That is correct, but wouldn't that also apply for the global user_agent option? If i have understood correctly the user_agent option will have the same "unwanted" affect. If we want to fix this problem, shouldn't we also limit the global user_agent option to PHP_INI_SYSTEM? (Which will be hard due to a quite hard BC break)

Anyway. I can gladly change to PHP_INI_SYSTEM if no one is against this change.

TysonAndre replied:
> > Obviously, for file_get_contents, there's no way to override the user agent, so it's useful that user_agent can be overridden at runtime. But curl_setopt can already override the user agent
>
> Sorry for the late response. As far as i have checked you can override the user_agent during runtime using file_get_contents. Therefore i wanted to keep curl.user_agent in the same mode as the original user_agent.

Didn't realize that - I think the vast majority of calls to file_get_contents don't typically use a context, and curl is typically preferred (easier to set timeouts(lookup/request/connect), redirect behavior, allowed protocols, etc).

Also, in general, php aims for consistency within a module(aka php extension) (e.g. `ext/curl`) (individual modules do/may differ greatly from each other for naming, class/function usage, etc), which is why I personally strongly prefer PHP_INI_SYSTEM

ramsey commented:
@TysonAndre, thanks for the clarification regarding an RFC.

@paresy, as Tyson said, it's time to put together an RFC. The link to the "RFC howto" is earlier in this thread. If you'd like some help, I'm happy to provide some assistance.

paresy commented:
Thanks! I will create the RFC soon. It's time to request RFC karma.

paresy commented:
I resolved the php.ini related comments and published an RFC: https://wiki.php.net/rfc/curl_user_agent

cmb69 commented on ext/curl/tests/curl_user_agent.phpt:
```diff
@@ -0,0 +1,56 @@
+Test curl.user_agent ini entry
+--INI--
+curl.user_agent=test/1.0
+--SKIPIF--
+<?php include 'skipif.inc'; ?>
```
This need to be changed now;
```suggestion
--EXTENSIONS--
curl
```

paresy replied:
Thanks!

iluuu1994 commented:
@paresy Are you still planning on pursuing this RFC?

paresy commented:
Last time I asked on the dev mailing list the responses were very mixed and I am still unsure in which direction I should proceed. If someone could guide me a little bit I would try to get this RFC forward and into the voting phase.
